src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java

```diff
@@ -472,8 +522,172 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
+        transport.handleResponse(transport.capturedRequests()[0].requestId, TransportResponse.Empty.INSTANCE);
+        assertIndexShardCounter(1);
+        transport.clear();
+        request = new Request(shardId).timeout("100ms");
+        primaryPhase = action.new PrimaryPhase(request, listener);
+        primaryPhase.run();
+        assertIndexShardCounter(2);
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        // try with failure response
+        transport.handleResponse(transport.capturedRequests()[0].requestId, new CorruptIndexException("simulated", (String) null));
+        assertIndexShardCounter(1);
+    }
+
+    @Test
+    public void testReplicasCounter() throws Exception {
+        final ShardId shardId = new ShardId("test", 0);
+        clusterService.setState(state(shardId.index().getName(), true,
+                ShardRoutingState.STARTED, ShardRoutingState.STARTED));
+        final IndexShardRoutingTable shardRoutingTable = clusterService.state().routingTable().index(shardId.index().getName()).shard(shardId.id());
+        final ShardRouting primaryShard = shardRoutingTable.primaryShard();
+        final ShardIterator shardIt = shardRoutingTable.shardsIt();
+        final Request request = new Request();
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        action = new ActionWithDelay(ImmutableSettings.EMPTY, "testActionWithExceptions", transportService, clusterService, threadPool);
+        final TransportShardReplicationOperationAction<Request, Request, Response>.InternalRequest internalRequest = action.new InternalRequest(request);
+        internalRequest.concreteIndex(shardId.index().name());
+        TransportShardReplicationOperationAction<Request, Request, Response>.ReplicationPhase replicationPhase =
+                action.new ReplicationPhase(shardIt, request,
+                        new Response(), new ClusterStateObserver(clusterService, logger),
+                        primaryShard, internalRequest, listener, new AtomicReference<IndexShard>());
+
+        assertThat(replicationPhase.totalShards(), equalTo(2));
+        assertThat(replicationPhase.pending(), equalTo(1));
+        replicationPhase.run();
+        final CapturingTransport.CapturedRequest[] capturedRequests = transport.capturedRequests();
+        assertThat(capturedRequests.length, equalTo(1));
+        final Action.ReplicaOperationTransportHandler replicaOperationTransportHandler = action.new ReplicaOperationTransportHandler();
+        Thread t = new Thread() {
+            public void run() {
+                try {
+                    replicaOperationTransportHandler.messageReceived(action.new ReplicaOperationRequest(shardId, new Request()), createTransportChannel());
+                } catch (Exception e) {
+                }
+            }
+        };
+        t.start();
+        // operation should be ongoing
+        awaitBusy(new Predicate<Object>() {
```

Comment:
I tried and found that the await for not null was wrong to begin with. we should wait for counter to be at 2 instead. I think this would be a little more tricky to accomplish with a notify because the notify would then would have to happen on increment?